sympy/integrals/prde.py

```diff
@@ -359,35 +379,381 @@ def prde_no_cancel_b_small(b, Q, n, DE):
+        # yn = Sum(djn*fjn, (j, 1, rn))
+        fn, An = param_rischDE(ba, bd, Qyn, DE)
+
+    fn = [Poly(fa.as_expr()/fd.as_expr(), DE.t, field=True)
+            for fa, fd in fn]
+
+    rn = len(fn)
+    m = len(An) - rn
+
+    # there would be a need to form a new equation
+    # just like Dy + fy = Sum(ci*qi, (i, 1, m))
+    # the equation is
+    # Dh + bh = Sum(ci*qi, (i, 1, m)) - Sum(djn*D(fjn*t**n) - b*fj*t**n)
+    # the below thing is wrong
+    for N in range(n, -1, -1):
+        s1 = 0
+        for j in range(rn):
+            djn = An[m + j]
+            fjn = fn[j]
+            s1 += djn*(derivation(fjn*DE.t**n, DE) - b*fjn*DE.t**n)
+
+
+def prde_cancel_exp(b, Q, n, DE):
+    """
+    Pg, 237
+    """
+    with DecrementLevel(DE):
+        # equation (7.23)
+        # Note: b != ba/bd
+        ba, bd = frac_in(b + n*derivation(DE.t, DE)/DE.t, DE.t)
+        Qyn = [(Poly(q.nth(n), DE.t), Poly(1, DE.t)) for q in Q]
+        fn, An = param_rischDE(ba, bd, Qyn, DE)
+
+    fn = [Poly(fa.as_expr()/fd.as_expr(), DE.t, field=True)
+            for fa, fd in fn]
+
+
+def param_poly_rischDE(a, b, q, n, DE):
+    """Polynomial solutions of a parametric Risch differential equation.
+
+    Given a derivation D in k[t], a, b in k[t] relatively prime, and q
+    = [q1, ..., qm] in k[t]^m, return h = [h1, ..., hr] in k[t]^r and
+    a matrix A with m + r columns and entries in Const(k) such that
+    a*Dp + b*p = Sum(ci*qi, (i, 1, m)) has a solution p of degree <= n
+    in k[t] with c1, ..., cm in Const(k) if and only if p = Sum(dj*hj,
+    (j, 1, r)) where d1, ..., dr are in Const(k) and (c1, ..., cm,
+    d1, ..., dr) is a solution of Ax == 0.
+    """
+    m = len(q)
+    if n < 0:
+        # Only the trivial zero solution is possible.
+        # Find relations between the qi.
+        if all([qi.is_zero for qi in q]):
+            return [], zeros(1, m)  # No constraints.
+
+        N = max([qi.degree(DE.t) for qi in q])
+        M = Matrix(N + 1, m, lambda i, j: q[j].nth(i))
+        A, _ = constant_system(M, zeros(M.rows, 1), DE)
+
+        return [], A
+
+    if a.is_ground:
+        # Normalization: a = 1.
+        a = a.LC()
+        b, q = b.quo_ground(a), [qi.quo_ground(a) for qi in q]
+
+        if not b.is_zero and (DE.case == 'base' or
+                b.degree(DE.t) > max(0, DE.d.degree(DE.t) - 1)):
+            return prde_no_cancel_b_large(b, q, n, DE)
+
+        elif ((b.is_zero or b.degree(DE.t) < DE.d.degree(DE.t) - 1)
+                and (DE.case == 'base' or DE.d.degree(DE.t) >= 2)):
+            return prde_no_cancel_b_small(b, q, n, DE)
+
+        elif (DE.d.degree(DE.t) >= 2 and
+              b.degree(DE.t) == DE.d.degree(DE.t) - 1 and
+              n > -b.as_poly(DE.t).LC()/DE.d.as_poly(DE.t).LC()):
+            raise NotImplementedError("prde_no_cancel_b_equal() is "
+                "not yet implemented.")
+        else:
+            if b.is_zero:
+                pass
```

Comment:
`a.LC()` may not be optimal here since it sympifies the coefficient. `a.nth(0)` does the same. 
But it does not matter since `quo_ground` will convert `a` back anyway. (Only some time is lost in the conversions.) And `a` should not be used elsewhere.